servers/audio/effects/audio_effect_limiter.cpp

```diff
@@ -29,38 +29,65 @@
 /*************************************************************************/
 
 #include "audio_effect_limiter.h"
-
+#include "core/math/math_funcs.h"
+
+// Soft clipping starts to reduce the peaks a little below the threshold level
+// and progressively increases its effect as the input level increases such that
+// the threshold is never exceeded.
+//
+// This is implemented as a logistics function because it's designed to increase near
+// linearly and taper up to a certain max value. This is exactly what we want for
+// this soft clipping function.
 void AudioEffectLimiterInstance::process(const AudioFrame *p_src_frames, AudioFrame *p_dst_frames, int p_frame_count) {
-	float threshdb = base->threshold;
+	// The threshold from which the limiter begins to be active, in decibels. Value can range from -30 to 0.
+	float threshold = Math::db2linear(base->threshold);
+
+	// The waveform's maximum allowed value, in decibels. Value can range from -20 to -0.1.
 	float ceiling = Math::db2linear(base->ceiling);
-	float ceildb = base->ceiling;
-	float makeup = Math::db2linear(ceildb - threshdb);
-	float sc = -base->soft_clip;
-	float scv = Math::db2linear(sc);
-	float peakdb = ceildb + 25;
-	float scmult = Math::abs((ceildb - sc) / (peakdb - sc));
+
+	// Applies a gain to the limited waves, in decibels. Value can range from 0 to 6.
+	// This will be the slope (k) of the logistics curve. 0 will end up creating a hard clip.
+	float soft_clip = Math::db2linear(base->soft_clip);
+
+	// The difference between the threshold and the ceiling
+	// This is used as the maximum value (L) in the logistics function.
+	float limit = ceiling - threshold;
+
+	// The halfway point of the threshold and the ceiling.
+	// This will serve as our x0 in the logistics transform.
+	float mid_point = limit / 2.0;
 
 	for (int i = 0; i < p_frame_count; i++) {
 		float spl0 = p_src_frames[i].l;
 		float spl1 = p_src_frames[i].r;
-		spl0 = spl0 * makeup;
-		spl1 = spl1 * makeup;
+
+		// The signs of the frame.
 		float sign0 = (spl0 < 0.0 ? -1.0 : 1.0);
 		float sign1 = (spl1 < 0.0 ? -1.0 : 1.0);
+
+		// The values of the frame.
 		float abs0 = Math::abs(spl0);
 		float abs1 = Math::abs(spl1);
-		float overdb0 = Math::linear2db(abs0) - ceildb;
-		float overdb1 = Math::linear2db(abs1) - ceildb;
 
-		if (abs0 > scv) {
-			spl0 = sign0 * (scv + Math::db2linear(overdb0 * scmult));
-		}
-		if (abs1 > scv) {
-			spl1 = sign1 * (scv + Math::db2linear(overdb1 * scmult));
+		// We only activate when we hit the threshold.
+		if (abs0 > threshold) {
+			// Pull the value down to the threshold.
+			abs0 -= threshold;
+			// Apply the logistics transformation.
+			abs0 = limit / (1 + exp(-soft_clip * (abs0 - mid_point)));
```

Comment:
What happens here when `soft_clip = 0` and `ceiling != threshold`? This would introduce discontinuities I think. Is there any way we could special case this, or somehow make the user experience a little bit nicer? Especially since the property hints prevent the user from setting both to 0.0.

Reply:
I actually did look at this and discovered that I misused the logistic function in my original implementation, leading to strange audio behavior that I didn't test for.

The Limit should be the Ceiling and the Midpoint should be threshold.

Reply:
Glad I asked! @ tag me in a comment when it's ready for another look and do some testing.

Reply:
One more thing, if you decide that the property hints should be different, definitely change them appropriately. For a user wanting a limiter that's completely linear up to a hard clip, I _think_ that would mean `ceiling` and `threshold` should both be set to 0.0 so those hints may need to change, but I'm not really sure. We definitely want to support linear behavior up to a hard clip though, but being able to blend in a little bit of softness in a configurable way will let people have a lot of control which is wonderful.